In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/nowplayingrs/sentiment_values.csv
/kaggle/input/nowplayingrs/user_track_hashtag_timestamp.csv
/kaggle/input/nowplayingrs/context_content_features.csv


## OBSERVER & SCRUB: Understand the Datasets and Data Preparation

* Data type conversions (e.g. numeric data mistakenly encoded as objects)
* Detect and deal with missing values
* Remove unnecessary columns

In [2]:
df1 = pd.read_csv('../input/nowplayingrs/sentiment_values.csv')

#Look at size of the dataset
df1.shape

(5290, 17)

In [3]:
#Look at the columns and initial rows of the dataset
df1.head()

hashtag   vader_min   vader_max   vader_sum  \
relaxtime         0.8 0.8 2.4      0.8         NaN         NaN         NaN   
melovechilicheese 0.8 0.8 0.8      0.8         NaN         NaN         NaN   
greatmusic        0.8 0.8 2.4      0.8         1.0         1.0         1.0   
rockballad        0.7 0.7 0.7      0.7         NaN         NaN         NaN   
amonamarth        0.3 0.3 0.3      0.3         NaN         NaN         NaN   

                               vader_avg    afinn_min   afinn_max   afinn_sum  \
relaxtime         0.8 0.8 2.4        NaN       0.7375      0.7375      0.7375   
melovechilicheese 0.8 0.8 0.8        NaN       0.9000      0.9000      0.9000   
greatmusic        0.8 0.8 2.4        1.0       0.8875      0.8875      0.8875   
rockballad        0.7 0.7 0.7        NaN          NaN         NaN         NaN   
amonamarth        0.3 0.3 0.3        NaN          NaN         NaN         NaN   

                                afinn_avg   ol_min   ol_max   ol_sum   ol_avg  \
relaxtime         0.8 0.8 2.4      0.7375      NaN      NaN      NaN      NaN   
melovechilicheese 0.8 0.8 0.8      0.9000      1.0      1.0      1.0      1.0   
greatmusic        0.8 0.8 2.4      0.8875      1.0      1.0      1.0      1.0   
rockballad        0.7 0.7 0.7         NaN      NaN      NaN      NaN      NaN   
amonamarth        0.3 0.3 0.3         NaN      0.0      0.0      0.0      0.0   

                                ss_min   ss_max   ss_sum   ss_avg  
relaxtime         0.8 0.8 2.4      NaN      NaN      NaN      NaN  
melovechilicheese 0.8 0.8 0.8      0.8      0.8      0.8      0.8  
greatmusic        0.8 0.8 2.4      0.8      0.8      0.8      0.8  
rockballad        0.7 0.7 0.7      NaN      NaN      NaN      NaN  
amonamarth        0.3 0.3 0.3      NaN      NaN      NaN      NaN

In [4]:
#Rename hashtag column
df1.rename(columns = {'hashtag':'ss_score'}, inplace = True)

In [5]:
##### Reset index
df1.reset_index(inplace=True)

#Rename columns
df1.rename(columns = {'level_0':'hashtag','level_1':'vader_score','level_2':'afinn_score','level_3':'ol_score'}, inplace = True)

#Show dataset to confirm changes
df1.head()

hashtag  vader_score  afinn_score  ol_score  ss_score  \
0          relaxtime          0.8          0.8       2.4       0.8   
1  melovechilicheese          0.8          0.8       0.8       0.8   
2         greatmusic          0.8          0.8       2.4       0.8   
3         rockballad          0.7          0.7       0.7       0.7   
4         amonamarth          0.3          0.3       0.3       0.3   

    vader_min   vader_max   vader_sum  vader_avg    afinn_min  ...  \
0         NaN         NaN         NaN        NaN       0.7375  ...   
1         NaN         NaN         NaN        NaN       0.9000  ...   
2         1.0         1.0         1.0        1.0       0.8875  ...   
3         NaN         NaN         NaN        NaN          NaN  ...   
4         NaN         NaN         NaN        NaN          NaN  ...   

    afinn_sum   afinn_avg   ol_min   ol_max   ol_sum   ol_avg   ss_min  \
0      0.7375      0.7375      NaN      NaN      NaN      NaN      NaN   
1      0.9000      0.9000      1.0      1.0      1.0      1.0      0.8   
2      0.8875      0.8875      1.0      1.0      1.0      1.0      0.8   
3         NaN         NaN      NaN      NaN      NaN      NaN      NaN   
4         NaN         NaN      0.0      0.0      0.0      0.0      NaN   

    ss_max   ss_sum   ss_avg  
0      NaN      NaN      NaN  
1      0.8      0.8      0.8  
2      0.8      0.8      0.8  
3      NaN      NaN      NaN  
4      NaN      NaN      NaN  

[5 rows x 21 columns]

In [6]:
#Set columns
df1.columns=['hashtag','vader_score','afinn_score','ol_score','ss_score','vader_min','vader_max','vader_sum','vader_avg','afinn_min',
            'afinn_max','afinn_sum','afinn_avg','ol_min','ol_max','ol_sum','ol_avg','ss_min','ss_max','ss_sum','ss_avg']

The following 12 columns are unecessary and will be removed from the data set:

- **vader_min**: no valuable information, will use vader_score instead
- **vader_max**: no valuable information, will use vader_score instead
- **vader_sum**: no valuable information, will use vader_score instead
- **affin_min**: no valuable information, will use affin_score instead
- **affin_max**: no valuable information, will use affin_score instead
- **affin_sum**: no valuable information, will use affin_score instead
- **ol_min**: no valuable information, will use ol_score instead
- **ol_max**: no valuable information, will use ol_score instead
- **ol_sum**: no valuable information, will use ol_score instead
- **ss_min**: no valuable information, will use ss_score instead
- **ss_max**: no valuable information, will use ss_score instead
- **ss_sum**: no valuable information, will use ss_score instead

In [7]:
df_tweets = df1.drop(['vader_min','vader_max','vader_sum','afinn_min','afinn_max','afinn_sum','ol_min','ol_max','ol_sum','ss_min','ss_max','ss_sum'], axis=1)
df_tweets.head()

hashtag  vader_score  afinn_score  ol_score  ss_score  vader_avg  \
0          relaxtime          0.8          0.8       2.4       0.8        NaN   
1  melovechilicheese          0.8          0.8       0.8       0.8        NaN   
2         greatmusic          0.8          0.8       2.4       0.8        1.0   
3         rockballad          0.7          0.7       0.7       0.7        NaN   
4         amonamarth          0.3          0.3       0.3       0.3        NaN   

   afinn_avg  ol_avg  ss_avg  
0     0.7375     NaN     NaN  
1     0.9000     1.0     0.8  
2     0.8875     1.0     0.8  
3        NaN     NaN     NaN  
4        NaN     0.0     NaN

In [8]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 9 columns):
hashtag        5290 non-null object
vader_score    3867 non-null float64
afinn_score    3867 non-null float64
ol_score       3867 non-null float64
ss_score       3867 non-null float64
vader_avg      255 non-null float64
afinn_avg      2635 non-null float64
ol_avg         2823 non-null float64
ss_avg         2160 non-null float64
dtypes: float64(8), object(1)
memory usage: 372.1+ KB


In [9]:
#Fill in missing vader_score with vader_avg score, if available
df_tweets['vader_score'] = df_tweets.apply(
    lambda row: row['vader_avg'] if np.isnan(row['vader_score']) else row['vader_score'],
    axis=1
)

In [10]:
#vader_score didn't change
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 9 columns):
hashtag        5290 non-null object
vader_score    3871 non-null float64
afinn_score    3867 non-null float64
ol_score       3867 non-null float64
ss_score       3867 non-null float64
vader_avg      255 non-null float64
afinn_avg      2635 non-null float64
ol_avg         2823 non-null float64
ss_avg         2160 non-null float64
dtypes: float64(8), object(1)
memory usage: 372.1+ KB


In [11]:
#Fill in missing afinn_score with afinn_avg score, if available
df_tweets['afinn_score'] = df_tweets.apply(
    lambda row: row['afinn_avg'] if np.isnan(row['afinn_score']) else row['afinn_score'],
    axis=1
)

In [12]:
#afinn_score increased from 3867 to 4532
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 9 columns):
hashtag        5290 non-null object
vader_score    3871 non-null float64
afinn_score    4532 non-null float64
ol_score       3867 non-null float64
ss_score       3867 non-null float64
vader_avg      255 non-null float64
afinn_avg      2635 non-null float64
ol_avg         2823 non-null float64
ss_avg         2160 non-null float64
dtypes: float64(8), object(1)
memory usage: 372.1+ KB


In [13]:
#Fill in missing ol_score with ol_avg score, if available
df_tweets['ol_score'] = df_tweets.apply(
    lambda row: row['ol_avg'] if np.isnan(row['ol_score']) else row['ol_score'],
    axis=1
)

In [14]:
#ol_score increased from 3867 to 4831
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 9 columns):
hashtag        5290 non-null object
vader_score    3871 non-null float64
afinn_score    4532 non-null float64
ol_score       4831 non-null float64
ss_score       3867 non-null float64
vader_avg      255 non-null float64
afinn_avg      2635 non-null float64
ol_avg         2823 non-null float64
ss_avg         2160 non-null float64
dtypes: float64(8), object(1)
memory usage: 372.1+ KB


In [15]:
#Fill in missing ss_score with ss_avg score, if available
df_tweets['ss_score'] = df_tweets.apply(
    lambda row: row['ss_avg'] if np.isnan(row['ss_score']) else row['ss_score'],
    axis=1
)

In [16]:
#ss_score increased from 3867 to 4471
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 9 columns):
hashtag        5290 non-null object
vader_score    3871 non-null float64
afinn_score    4532 non-null float64
ol_score       4831 non-null float64
ss_score       4471 non-null float64
vader_avg      255 non-null float64
afinn_avg      2635 non-null float64
ol_avg         2823 non-null float64
ss_avg         2160 non-null float64
dtypes: float64(8), object(1)
memory usage: 372.1+ KB


In [17]:
#Remove all of the unnecessary scores - ol_score has the highest amount of ratings per hashtag
df_tweets = df_tweets.drop(['vader_score','afinn_score','ss_score','vader_avg','afinn_avg','ol_avg','ss_avg'], axis=1)
df_tweets.head()

hashtag  ol_score
0          relaxtime       2.4
1  melovechilicheese       0.8
2         greatmusic       2.4
3         rockballad       0.7
4         amonamarth       0.3

In [18]:
#Load second dataset
df2 = pd.read_csv('../input/nowplayingrs/user_track_hashtag_timestamp.csv')

#Look at size of the dataset
df2.shape

(17560113, 4)

In [19]:
#Look at the columns and initial rows of the dataset
df2.head()

user_id                          track_id     hashtag  \
0    81496937  cd52b3e5b51da29e5893dba82a418a4b  nowplaying   
1    81496937  cd52b3e5b51da29e5893dba82a418a4b        goth   
2    81496937  cd52b3e5b51da29e5893dba82a418a4b   deathrock   
3    81496937  cd52b3e5b51da29e5893dba82a418a4b    postpunk   
4  2205686924  da3110a77b724072b08f231c9d6f7534  NowPlaying   

            created_at  
0  2014-01-01 05:54:21  
1  2014-01-01 05:54:21  
2  2014-01-01 05:54:21  
3  2014-01-01 05:54:21  
4  2014-01-01 05:54:22

In [20]:
#Check for null values
df2.apply(lambda x: x.isnull().sum())

user_id       0
track_id      0
hashtag       1
created_at    0
dtype: int64

In [21]:
#Drop null row
df2.dropna(subset=['hashtag'], inplace=True)
df2.apply(lambda x: x.isnull().sum())

user_id       0
track_id      0
hashtag       0
created_at    0
dtype: int64

In [22]:
# Get the count of the track_id
counts = df2['track_id'].value_counts()

# Select the items where the track_id count is less than 10 and remove them
df2 = df2[~df2['track_id'].isin(counts[counts < 10].index)]

# Show info
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16873184 entries, 0 to 17560112
Data columns (total 4 columns):
user_id       int64
track_id      object
hashtag       object
created_at    object
dtypes: int64(1), object(3)
memory usage: 643.7+ MB


In [23]:
#Merge CSV files into a single file based on hashtag
df_sentiment = pd.merge(df_tweets, df2, on="hashtag", how='inner')
df_sentiment.head()

hashtag  ol_score     user_id                          track_id  \
0          relaxtime       2.4   104415394  9175ac1532ee7dbe97602866efabac58   
1          relaxtime       2.4   240771401  c2da30eb3450e8a3e5bfa16e8fa527da   
2          relaxtime       2.4   637256774  dcbb5aff8f96a79be9f59bc0e7b5c38d   
3  melovechilicheese       0.8  2438833016  76ed33b994cc15575614b91284b965bb   
4         greatmusic       2.4   252330820  8f2ac86abb8bd48273c8fc95b632e347   

            created_at  
0  2014-05-20 07:24:40  
1  2014-10-12 22:46:57  
2  2014-11-07 13:01:08  
3  2014-05-05 23:58:21  
4  2014-02-13 16:18:51

In [24]:
#Confirm null values in new dataframe
df_sentiment.apply(lambda x: x.isnull().sum())

hashtag           0
ol_score      69714
user_id           0
track_id          0
created_at        0
dtype: int64

In [25]:
#Fill all null values for ol_score column with zeros
df_sentiment['ol_score'].fillna(0, inplace=True)
df_sentiment.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5873847 entries, 0 to 5873846
Data columns (total 5 columns):
hashtag       object
ol_score      float64
user_id       int64
track_id      object
created_at    object
dtypes: float64(1), int64(1), object(3)
memory usage: 268.9+ MB


Now that the two CSV files are joined, due to the inner join, the new dataframe `df_join` is reduced to **5,873,847** rows (from 17,560,114).

In [ ]:
#df_sentiment.to_csv('../input/nowplayingrs/sentiment_cleaned.csv', index=False)

In [27]:
#Load third dataset and limit it to only load 22 columns
df3 = pd.read_csv('../input/nowplayingrs/context_content_features.csv', usecols=range(0, 22))

#Look at size of the dataset
df3.shape

(11614671, 22)

In [28]:
# Get the count of the track_id
counts = df3['track_id'].value_counts()

# Select the items where the track_id count is less than 10 and remove them
df3 = df3[~df3['track_id'].isin(counts[counts < 10].index)]

# Show info
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10887911 entries, 0 to 11614670
Data columns (total 22 columns):
coordinates         object
instrumentalness    float64
liveness            float64
speechiness         float64
danceability        float64
valence             float64
loudness            float64
tempo               float64
acousticness        float64
energy              float64
mode                float64
key                 float64
artist_id           object
place               object
geo                 object
tweet_lang          object
track_id            object
created_at          object
lang                object
time_zone           object
user_id             float64
id                  int64
dtypes: float64(12), int64(1), object(9)
memory usage: 1.9+ GB


In [29]:
#Drop unnecessary/duplicate columns before merging with df_join dataframe
df3 = df3.drop(['user_id','coordinates','id','place','geo'], axis=1)
df3.head()

instrumentalness  liveness  speechiness  danceability  valence  loudness  \
0           0.00479    0.1800       0.0294         0.634    0.342    -8.345   
1           0.01770    0.0638       0.0624         0.769    0.752    -8.252   
2           0.00000    0.0860       0.0436         0.675    0.775    -4.432   
3           0.00000    0.1430       0.0292         0.324    0.333    -5.647   
5           0.00000    0.1100       0.0375         0.641    0.912    -4.271   

     tempo  acousticness  energy  mode  key                         artist_id  \
0  125.044       0.00035   0.697   1.0  6.0  b2980c722a1ace7a30303718ce5491d8   
1   95.862       0.26700   0.826   1.0  7.0  5cddcd0e314e2f2223ab21937d2c8778   
2   97.030       0.21700   0.885   0.0  1.0  e41273f43af504714d85465294f1f369   
3   74.101       0.23900   0.574   1.0  7.0  557ce373bd29743eb00a3723ab19ebe8   
5   93.010       0.02680   0.787   1.0  0.0  f965ec352eb8c0efc0af46244754942f   

  tweet_lang                          track_id           created_at lang  \
0         en  cd52b3e5b51da29e5893dba82a418a4b  2014-01-01 05:54:21   en   
1         en  da3110a77b724072b08f231c9d6f7534  2014-01-01 05:54:22   en   
2         en  ba84d88c10fb0e42d4754a27ead10546  2014-01-01 05:54:22   es   
3         en  33f95122281f76e7134f9cbea3be980f  2014-01-01 05:54:24   en   
5         en  8bd5206b84c968eda0af8bc86d6ab1d1  2014-01-01 05:54:25   en   

                     time_zone  
0   Central Time (US & Canada)  
1                          NaN  
2  Mountain Time (US & Canada)  
3   Eastern Time (US & Canada)  
5   Central Time (US & Canada)

In [30]:
#Convert created_at column to datetime
df3['created_at'] = pd.to_datetime(df3['created_at'])
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10887911 entries, 0 to 11614670
Data columns (total 17 columns):
instrumentalness    float64
liveness            float64
speechiness         float64
danceability        float64
valence             float64
loudness            float64
tempo               float64
acousticness        float64
energy              float64
mode                float64
key                 float64
artist_id           object
tweet_lang          object
track_id            object
created_at          datetime64[ns]
lang                object
time_zone           object
dtypes: datetime64[ns](1), float64(11), object(5)
memory usage: 1.5+ GB


In [31]:
#Rename created_at column
df3.rename(columns = {'created_at':'timestamp'}, inplace = True)

In [32]:
#Drop all null value rows
df3 = df3.dropna()
df3.shape

(7781780, 17)

Reduced the dataset from 10,887,911 to **7,781,780** by dropping all null value rows

In [ ]:
#df3.to_csv('../input/nowplayingrs/features_cleaned.csv', index=False)

In [33]:
import dask.dataframe as dd

#Merge df_sentiment and df3 CSV files into new CSV file based on track_id
df4 = dd.merge(df_sentiment, df3, on="track_id", how='inner')
df4.head()

MemoryError: Unable to allocate array with shape (4826764770,) and data type int64

In [ ]:
#Create new dataframe based on datetime range for training set
df_training = df4.loc['2014-01-01':'2014-09-30']
df_training.shape

In [ ]:
#Create new dataframe based on datetime range for test set
df_test = df4.loc['2014-11-01':'2014-12-23']
df_test.shape